GoogleDataTransport/GDTCCTLibrary/GDTCCTUploader.m

```diff
@@ -190,210 +206,341 @@ - (NSString *)FLLAndCSHAPIKey {
 #
 
 - (void)uploadTarget:(GDTCORTarget)target withConditions:(GDTCORUploadConditions)conditions {
-  __block GDTCORBackgroundIdentifier bgID = GDTCORBackgroundIdentifierInvalid;
-  bgID = [[GDTCORApplication sharedApplication]
+  __block GDTCORBackgroundIdentifier backgroundTaskID = GDTCORBackgroundIdentifierInvalid;
+
+  dispatch_block_t backgroundTaskCompletion = ^{
+    // End the background task if there was one.
+    if (backgroundTaskID != GDTCORBackgroundIdentifierInvalid) {
+      [[GDTCORApplication sharedApplication] endBackgroundTask:backgroundTaskID];
+      backgroundTaskID = GDTCORBackgroundIdentifierInvalid;
+    }
+  };
+
+  backgroundTaskID = [[GDTCORApplication sharedApplication]
       beginBackgroundTaskWithName:@"GDTCCTUploader-upload"
                 expirationHandler:^{
-                  if (bgID != GDTCORBackgroundIdentifierInvalid) {
+                  if (backgroundTaskID != GDTCORBackgroundIdentifierInvalid) {
                     // Cancel the upload and complete delivery.
                     [self.currentTask cancel];
 
                     // End the background task.
-                    [[GDTCORApplication sharedApplication] endBackgroundTask:bgID];
+                    backgroundTaskCompletion();
                   }
                 }];
 
   dispatch_async(_uploaderQueue, ^{
     id<GDTCORStorageProtocol> storage = GDTCORStorageInstanceForTarget(target);
 
-    if (![self readyToUploadTarget:target conditions:conditions]) {
-      return;
-    }
-    __block NSNumber *batchID;
+    // 1. Fetch events to upload.
+    [self batchToUploadForTarget:target
+                         storage:storage
+                      conditions:conditions
+                      completion:^(NSNumber *_Nullable batchID,
+                                   NSSet<GDTCOREvent *> *_Nullable events) {
+                        // 2. Check if there are events to upload.
+                        if (!events || events.count == 0) {
+                          dispatch_async(self.uploaderQueue, ^{
+                            GDTCORLogDebug(@"Target %ld reported as ready for upload, but no "
+                                           @"events were selected",
+                                           (long)target);
+                            self.isCurrentlyUploading = NO;
+                            backgroundTaskCompletion();
+                          });
+
+                          return;
+                        }
+                        // 3. Upload events.
+                        [self uploadBatchWithID:batchID
+                                         events:events
+                                         target:target
+                                        storage:storage
+                                     completion:^{
+                                       backgroundTaskCompletion();
```

Comment:
Yeah, it's handled in the method itself. Agree, the readability and using `isCurrentlyUploading` approach in  can be improved in general, but it will require bigger changes which I would prefer to implement in followup PRs.

Reply:
I agree. It took sometime to understand how this state property is managed.

Not sure if this could be a possible step forward here: can we have `isCurrentlyUploading = YES` in the step 3 above. This would atleast reduce to confusion of hopping different areas of the code to see how this state is managed. Just a nit!

Reply:
Unfortunately setting `isCurrentlyUploading = YES` in the step 3 is already a bit late and may lead to race condition. 

In the future I would like to refactor the class to use Promises that will allow to express the sequence more elaborative and will provide a way to manage the state i a single place. It is expected to be updated in the scope of b/160018519.